pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -791,53 +772,299 @@ func (r *MachineSetSyncReconciler) createOrUpdateCAPIInfraMachineTemplate(ctx co
+	}
+
+	return true, nil
+}
+
+// ensureMAPIMachineSetStatusUpdated updates the MAPI machine set status if changes are detected and conditions are met.
+func (r *MachineSetSyncReconciler) ensureMAPIMachineSetStatusUpdated(ctx context.Context, existingMAPIMachineSet *machinev1beta1.MachineSet, convertedMAPIMachineSet *machinev1beta1.MachineSet, updatedMAPIMachineSet *machinev1beta1.MachineSet, sourceCAPIMachineSet *clusterv1.MachineSet, mapiMachineSetsDiff map[string]any, specUpdated bool) (bool, error) {
+	logger := log.FromContext(ctx)
+
+	// If there are no status changes and the spec has not been updated, return early.
+	if !hasStatusChanges(mapiMachineSetsDiff) && !specUpdated {
+		return false, nil
+	}
+
+	// If the source API object (CAPI MachineSet) status.observedGeneration does not match the objectmeta.generation
+	// it means the source API object status has not yet caught up with the desired spec,
+	// so we don't want to update the MAPI machine set status until that has happened.
+	if sourceCAPIMachineSet.Status.ObservedGeneration != sourceCAPIMachineSet.ObjectMeta.Generation {
+		logger.Info("Changes detected for MAPI machine set status, but the CAPI machine spec has not been observed yet, skipping status update")
+		return false, nil
+	}
+
+	logger.Info("Changes detected for MAPI machine set status. Updating it", "diff", fmt.Sprintf("%+v", mapiMachineSetsDiff))
+
+	patchBase := client.MergeFrom(existingMAPIMachineSet.DeepCopy())
+
+	// Set the changed MAPI machine set status fields from the converted MAPI machine set.
+	setChangedMAPIMachineSetStatusFields(existingMAPIMachineSet, convertedMAPIMachineSet)
+
+	// Update the observed generation to match the updated source API object generation.
+	existingMAPIMachineSet.Status.ObservedGeneration = updatedMAPIMachineSet.ObjectMeta.Generation
+
+	isPatchRequired, err := util.IsPatchRequired(existingMAPIMachineSet, patchBase)
+	if err != nil {
+		return false, fmt.Errorf("failed to check if patch is required: %w", err)
+	}
+
+	if isPatchRequired {
+		if err := r.Status().Patch(ctx, existingMAPIMachineSet, patchBase); err != nil {
+			logger.Error(err, "Failed to update MAPI machine set status")
+
+			updateErr := fmt.Errorf("failed to update MAPI machine set status: %w", err)
+
+			if condErr := r.applySynchronizedConditionWithPatch(
+				ctx, existingMAPIMachineSet, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachineSet, updateErr.Error(), nil); condErr != nil {
+				return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+			}
+
+			return false, updateErr
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// setChangedCAPIMachineSetStatusFields sets the updated fields in the CAPI machine set status.
+func setChangedCAPIMachineSetStatusFields(existingCAPIMachineSet, convertedCAPIMachineSet *clusterv1.MachineSet) {
+	// convertedCAPIMachineSet holds the computed and desired status changes, so apply them to the existing existingCAPIMachineSet.
```

Comment:
Note: almost feels like we implement the conversion twice: MAPi -> CAPI -> CAPI.

(but let's keep as is for now I'd say)

There's a pattern of using an "change thing" func, which could be used on an existing or new object to only have on function modifying or doing stuff.

Reply:
This is only a renaming, the underlying pattern is one we already use in other places for this controller, so I've reused it for consistency, but we can change this in follow-up refactors if we feel like it.